Readme.md
# SCRIPT.ARTWORK.DOWNLOADER - Custom API Support

![icon](https://raw.githubusercontent.com/mgonzales71/script.artwork.downloader/custom-api/icon.png)
![fanart](https://raw.githubusercontent.com/mgonzales71/script.artwork.downloader/custom-api/fanart.jpg)

**Configure your Custom API Keys in Addon -> Settings**



This script can automaticly download all available artwork for your movies/tvshows in your library.
!!Each Movies and TV Show must have it's own subfolder!!

----------------------------------------------------------------------------------------------
To run the script in bulk mode simply click on it in the 'Addons' or 'Programs' menu in XBMC.
You can change the settings if you wish in the addon settings page.

This script has some different run modes. Please check the XBMC wiki page:

http://wiki.xbmc.org/index.php?title=Add-on:Artwork_Downloader


changelog.txt
v13.1.0.api
- Custom API Support Added

v12.0.29
- Update to fanart.tv API v3

v12.0.28
- Don't set simplejson module to optional

v12.0.27
- Fix wrong order of listing using votes for artwork from fanart.tv

v12.0.26
- Call TMDb config to get up-to-date image base URL

v12.0.25
- Always apply language filters else you end up with a random language

v12.0.24
- Fix unicode error in season query

v12.0.23
- Fix multiple text search for disctype in filename and avoiding ac3 dubs to be recognized as 3d (ex. ac3d)

v12.0.22
- Fix that some files where never found or downloaded

v12.0.21
- Fix change image url for TheMovieDB

v12.0.20
- Fix local scraping

v12.0.19
- Gracefully fail if a non library item is send to the script
- Fix not downloading extrafanart any more
- Correct local scraper behaviour
- Catch edge cases where it can't get data from library
- Fix skin calling methods

v12.0.18
- Fixed wrong variable

v12.0.17
- Fixed wrong filename for season landscape
- Fixed extra slash for BDMV and VIDEO_TS

v12.0.16
- A lot of code cleanup
- Updated language files from Transifex

v12.0.15
- Fixed custom mode
- Added option to show image selection directly if type is specified

v12.0.14
- Fixed missing brackets

v12.0.13
- Add better support for local artwork: allow selection in gui mode and also scrape local artwork without a valid media id
- Update language files

v12.0.12
- Update languages
- Fix script error on some season artwork

v12.0.10
- Bump dependency versions to prevent being marked as broken

v12.0.10
- Fixed correct check on season artwork so it isn't downloaded again and again
- Added quick hack to support local artwork if nothing present on fanart.tv

v12.0.9
- Fix wrong filename for season landscape
- Remove old code

v12.0.7
- Change API url

v12.0.6
- Fix report on wrong setting combination
- Enable some artwork by default
- Change poster and banner filenaming to match Frodo

v12.0.5
- Force overwrite when using GUI mode

v12.0.4
- Directly add the image URL to the databse
- Option to override with local file

v12.0.1
- Fix bulk downloading for a single movie/tvshow

v12.0.0
- XBMC12.0 Frodo version
- use ListItem.DBID for matching media items
- Depricate use of ListItem.Title and ListItem.Path
- Remove db caching because it's not needed anymore

v1.0.19
- Add fix (again) for Frodo users

v1.0.18
- Add fix for Frodo users

v1.0.17
- Forgot to update settings.xml

v1.0.16
- Added: Movie Thumb
- Added: Movie Banner
- Added: HD artwork takes priority

v1.0.15
- Disable autorun
- Disable fanart deletion and reduce max number of image

v1.0.14
- Fixed: Possible empty library list when there was no season data

v1.0.13
- Fixed: Forgot to remove test line

v1.0.12
- Fixed: Didn't cache empty results
- Fixed: Should now also work for multipath://
- Changed: Improved coding in some parts

v1.0.11
- Fixed: Image selection wasn't working when there was a username/password in source

v1.0.10
- Added: Korean translation (thx piodio)
- Added: Debug option
- Change: Code handling

v1.0.9
- Changed: increase the API cache expire time to stop trashing the fanart.tv site
- Added: Only show artwork types in GUI mode that are avaiable
- Fixed: Unicode error in custom/gui mode search

v1.0.8
- Added: Increased speed by using commingcache for library JSON when using custom/gui mode (expires after 1 hour)
- Added: Greek translation (thx CutSickAss)
- Added: Spanish translation (thx PatOso)
- added: Portuguese translation (Brazil) (thx Wanilton)
- added: French translation (thx mikebzh44)
- Added: Localization of the download report
- Added: TV Show Season Posters
- Added: TV Show Season Banners
- Added: TV Show Season Thumb 16:9
- Added: Movie Clearlogo
- Added: Movie Clearart
- Added: Movie Discart
- Fixed: Don't run script twice when it was started on service time caused by the speed increase by using commoncache
- Changed: Switched to the new fanart.tv API
- Changed: increase caching of API calls because provider data doesn't change that (24->48 hours)

v1.0.7 
- Changed: Revert 1.0.6 because dependencies are now met. Meaning options from 1.0.5 are now active
- Fixed: Limit TMDB by rating not by number of votes

v1.0.6

- Fixed: Revert 1.0.5 because dependencies were not met yet

v1.0.5
- Added: Use commoncache module to decrease API load on providers
- Added: TV Show seasonbanners "seasonx-banner.jpg"

v1.0.4
- Fixed: After doing ID conversion it skipped that movie
- Fixed: Some mayor bug that no ones has noticed so far
- Fixed: When no exact title match was found it downloaded the entire library
- Fixed: Close progress dialog before opening the image selection window
- Changed: Reverted old change of using image ID from tmdb as filename. This is not available any more in new API v3
- Changed: Use the new TMDB API v3
- Changed: Use smaller image in GUI dialog for faster loading of images (when posible)
- Changed: You can now set preferred language. Will always use English as fallback
- Changed: Show yes/no dialog for faulty settings instead of looping
- Changed: Only start service when user is logged in (requires minimal XBMC beta2)
- Added: Polish translation (thx cienislaw)
- Added: Dutch translation (thx Martijn)
- Added: German translation (thx Suven)
- Added: Download report is accessible from the add-on settings window
- Added: Download report is accessible from the yes/no dialog when finished (not when background/silent/custom run)
- Added: Extra info to gui dialog (language, rating, votes, season and size)
- Added: Ratings for movie artwork 
- Added: Now tries to look up the 'tmdb ID' if no 'imdb ID' is found and retrieves artwork using that
- Added: Added musicvideo section that uses the title to search for artwork on TMDB
         Needs extra argument from skin (see runscript.txt)

v1.0.3
- Fixed: Error below limit check
- Fixed: Custom download for specific movie artwork only
- Fixed: Downloads artwork for selected movie when multiple exist with same name

lib/settings.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
#     Copyright (C) 2011-2014 Martijn Kaijser
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program. If not, see <http://www.gnu.org/licenses/>.
#

#import modules
import xbmc
import xbmcaddon
import lib.common
from lib.utils import dialog_msg, log

### get addon info
__addon__        = lib.common.__addon__
__localize__     = lib.common.__localize__

### General seetting variables
def get():
    setting = {'failcount':                0,     # Initial fail count
               'failthreshold':            3,     # Abbort when this many fails
               'xmlfailthreshold':         5,     # Abbort when this many fails
               'api_timedelay':            5000,  # in msec

               'centralize_enable':        __addon__.getSetting("centralize_enable")      == 'true',
               'centralfolder_movies':     __addon__.getSetting("centralfolder_movies"),
               'centralfolder_tvshows':    __addon__.getSetting("centralfolder_tvshows"),
               'background':               __addon__.getSetting("background")             == 'true',
               'notify':                   __addon__.getSetting("notify")                 == 'true',
               'service_startup':          __addon__.getSetting("service_startup")        == 'true',
               'service_startupdelay':     __addon__.getSetting("service_startupdelay"),
               'service_enable':           __addon__.getSetting("service_enable")         == 'true',
               'service_runtime':          __addon__.getSetting("service_runtime"),
               'files_overwrite':          __addon__.getSetting("files_overwrite")        == 'true',
               'files_local':              __addon__.getSetting("files_local")            == 'true',
               'xbmc_caching_enabled':     __addon__.getSetting("xbmc_caching_enabled")   == 'true',
               'debug_enabled':            __addon__.getSetting("debug_enabled")          == 'true',
               'service_startup':          False,
               'service_enable':           False,
               'fanart_api_key':           __addon__.getSetting("fanart_api_key"),
               'tmdb_api_key':             __addon__.getSetting("tmdb_api_key"),

               'movie_enable':             __addon__.getSetting("movie_enable")           == 'true',
               'movie_poster':             __addon__.getSetting("movie_poster")           == 'true',
               'movie_fanart':             __addon__.getSetting("movie_fanart")           == 'true',
               'movie_extrafanart':        __addon__.getSetting("movie_extrafanart")      == 'true',
               'movie_extrathumbs':        __addon__.getSetting("movie_extrathumbs")      == 'true',
               'movie_logo':               __addon__.getSetting("movie_logo")             == 'true',
               'movie_clearart':           __addon__.getSetting("movie_clearart")         == 'true',
               'movie_discart':            __addon__.getSetting("movie_discart")          == 'true',
               'movie_landscape':          __addon__.getSetting("movie_landscape")        == 'true',
               'movie_banner':             __addon__.getSetting("movie_banner")           == 'true',

               'tvshow_enable':            __addon__.getSetting("tvshow_enable")          == 'true',
               'tvshow_poster':            __addon__.getSetting("tvshow_poster")          == 'true',
               'tvshow_seasonposter':      __addon__.getSetting("tvshow_seasonposter")    == 'true',
               'tvshow_fanart':            __addon__.getSetting("tvshow_fanart")          == 'true',
               'tvshow_extrafanart':       __addon__.getSetting("tvshow_extrafanart")     == 'true',
               'tvshow_clearart':          __addon__.getSetting("tvshow_clearart")        == 'true',
               'tvshow_logo':              __addon__.getSetting("tvshow_logo")            == 'true',
               'tvshow_landscape':         __addon__.getSetting("tvshow_landscape")       == 'true',
               'tvshow_seasonlandscape':   __addon__.getSetting("tvshow_seasonlandscape") == 'true',
               'tvshow_showbanner':        __addon__.getSetting("tvshow_showbanner")      == 'true',
               'tvshow_seasonbanner':      __addon__.getSetting("tvshow_seasonbanner")    == 'true',
               'tvshow_characterart':      __addon__.getSetting("tvshow_characterart")    == 'true',

               'musicvideo_enable':        __addon__.getSetting("musicvideo_enable")     == 'true',
               'musicvideo_poster':        __addon__.getSetting("musicvideo_poster")     == 'true',
               'musicvideo_fanart':        __addon__.getSetting("musicvideo_fanart")     == 'true',
               'musicvideo_extrafanart':   __addon__.getSetting("musicvideo_extrafanart")== 'true',
               'musicvideo_extrathumbs':   __addon__.getSetting("musicvideo_extrathumbs")== 'true',
               'musicvideo_logo':          __addon__.getSetting("musicvideo_logo")       == 'true',
               'musicvideo_clearart':      __addon__.getSetting("musicvideo_clearart")   == 'true',
               'musicvideo_discart':       __addon__.getSetting("musicvideo_discart")    == 'true'}
    return setting

def get_limit():
    setting = {'limit_artwork':            __addon__.getSetting("limit_artwork")          == "true",
               'limit_extrafanart_max':    (float(__addon__.getSetting("limit_extrafanart_maximum"))),
               'limit_extrafanart_rating': int(float(__addon__.getSetting("limit_extrafanart_rating"))),
               'limit_size_moviefanart':   int(__addon__.getSetting("limit_size_moviefanart")),
               'limit_size_tvshowfanart':  int(__addon__.getSetting("limit_size_tvshowfanart")),
               'limit_extrathumbs':        True,
               'limit_extrathumbs_max':    4,
               'limit_artwork_max':        1,
               'limit_preferred_language': __addon__.getSetting("limit_preferred_language"),
               'limit_notext':             __addon__.getSetting("limit_notext")           == 'true'}
    return setting
    
### Check for faulty setting combinations
def check():
    setting = get()
    settings_faulty = True
    while settings_faulty:
        settings_faulty = True
        check_movie = check_tvshow = check_musicvideo = check_centralize = True
        # re-check settings after posible change
        setting = get()
        # Check if faulty setting in movie section
        if setting.get('movie_enable'):
            if not setting.get('movie_poster') and not setting.get('movie_fanart') and not setting.get('movie_extrafanart') and not setting.get('movie_extrathumbs') and not setting.get('movie_logo') and not setting.get('movie_clearart') and not setting.get('movie_discart') and not setting.get('movie_landscape') and not setting.get('movie_banner'):
                check_movie = False
                log('Setting check: No subsetting of movies enabled')
            else: check_movie = True
        # Check if faulty setting in tvshow section
        if setting.get('tvshow_enable'):
            if not setting.get('tvshow_poster') and not setting.get('tvshow_seasonposter') and not setting.get('tvshow_fanart') and not setting.get('tvshow_extrafanart') and not setting.get('tvshow_clearart') and not setting.get('tvshow_characterart') and not setting.get('tvshow_logo') and not setting.get('tvshow_showbanner') and not setting.get('tvshow_seasonbanner') and not setting.get('tvshow_landscape') and not setting.get('tvshow_seasonlandscape'):
                check_tvshow = False
                log('Setting check: No subsetting of tv shows enabled')
            else: check_tvshow = True
        # Check if faulty setting in musicvideo section
        if setting.get('musicvideo_enable'):
            if not setting.get('musicvideo_poster') and not setting.get('musicvideo_fanart') and not setting.get('musicvideo_extrafanart') and not setting.get('musicvideo_extrathumbs') and not setting.get('musicvideo_logo') and not setting.get('musicvideo_clearart') and not setting.get('musicvideo_discart'):
                check_musicvideo = False
                log('Setting check: No subsetting of musicvideo enabled')
            else: check_musicvideo = True
        # Check if faulty setting in centralize section
        if setting.get('centralize_enable'):
            if setting.get('centralfolder_movies') == '' and setting.get('centralfolder_tvshows') == '':
                check_centralize = False
                log('Setting check: No centralized folder chosen')
            else: check_centralize = True
        # Compare all setting check
        if check_movie and check_tvshow and check_musicvideo and check_centralize:
            settings_faulty = False
        else: settings_faulty = True
        # Faulty setting found
        if settings_faulty:
            log('Faulty setting combination found')
            # when faulty setting detected ask to open the settings window
            if dialog_msg('yesno', line1 = __localize__(32003), line2 = __localize__(32004), background = False, nolabel = __localize__(32026), yeslabel = __localize__(32025)):
                __addon__.openSettings()
            # if not cancel the script
            else:
                return False
        else:
            return True


lib/provider/tmdb.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
#     Copyright (C) 2011-2014 Martijn Kaijser
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program. If not, see <http://www.gnu.org/licenses/>.
#

#import modules
import xbmc
import sys

### import libraries
from lib.language import *
from lib.script_exceptions import NoFanartError
from lib.utils import *
from lib.settings import get
from operator import itemgetter

### get addon info
__localize__    = ( sys.modules[ "__main__" ].__localize__ )
setting = get()

API_KEY = '4be68d7eab1fbd1b6fd8a3b80a65a95e'
API_CFG = 'http://api.themoviedb.org/3/configuration?api_key=%s'
API_URL = 'http://api.themoviedb.org/3/movie/%s/images?api_key=%s'
TMDB_API_KEY = setting['tmdb_api_key']
if TMDB_API_KEY:
    API_KEY = TMDB_API_KEY
    API_CFG = API_CFG + "&tmdb_api_key=" + TMDB_API_KEY
    API_URL = API_URL + "&tmdb_api_key=" + TMDB_API_KEY

class TMDBProvider():

    def __init__(self):
        self.name = 'TMDB'

    def get_image_list(self, media_id):
        image_list = []
        api_cfg = get_data(API_CFG%(API_KEY), 'json')
        if api_cfg == "Empty" or not api_cfg:
            return image_list
        BASE_IMAGEURL = api_cfg['images'].get('base_url')
        data = get_data(API_URL%(media_id, API_KEY), 'json')
        if data == "Empty" or not data:
            return image_list
        else:
            # Get fanart
            try:
                for item in data['backdrops']:
                    if int(item.get('vote_count')) >= 1:
                        rating = float( "%.1f" % float( item.get('vote_average'))) #output string with one decimal
                        votes = item.get('vote_count','n/a')
                    else:
                        rating = 'n/a'
                        votes = 'n/a'
                    image_list.append({'url': BASE_IMAGEURL + 'original' + item['file_path'],
                                       'preview': BASE_IMAGEURL + 'w300' + item['file_path'],
                                       'id': item.get('file_path').lstrip('/').replace('.jpg', ''),
                                       'art_type': ['fanart','extrafanart'],
                                       'height': item.get('height'),
                                       'width': item.get('width'),
                                       'language': item.get('iso_639_1','n/a'),
                                       'rating': rating,
                                       'votes': votes,
                                       # Create Gui string to display
                                       'generalinfo': ('%s: %s  |  %s: %s  |  %s: %s  |  %s: %sx%s  |  ' 
                                                       %( __localize__(32141), get_language(item.get('iso_639_1','n/a')).capitalize(),
                                                          __localize__(32142), rating,
                                                          __localize__(32143), votes,
                                                          __localize__(32145), item.get('width'), item.get('height')))})
            except Exception, e:
                log( 'Problem report: %s' %str( e ), xbmc.LOGNOTICE )
            # Get thumbs
            try:
                for item in data['backdrops']:
                    if int(item.get('vote_count')) >= 1:
                        rating = float( "%.1f" % float( item.get('vote_average'))) #output string with one decimal
                        votes = item.get('vote_count','n/a')
                    else:
                        rating = 'n/a'
                        votes = 'n/a'
                    # Fill list
                    image_list.append({'url': BASE_IMAGEURL + 'w780' + item['file_path'],
                                       'preview': BASE_IMAGEURL + 'w300' + item['file_path'],
                                       'id': item.get('file_path').lstrip('/').replace('.jpg', ''),
                                       'art_type': ['extrathumbs'],
                                       'height': item.get('height'),
                                       'width': item.get('width'),
                                       'language': item.get('iso_639_1','n/a'),
                                       'rating': rating,
                                       'votes': votes,
                                       # Create Gui string to display
                                       'generalinfo': ('%s: %s  |  %s: %s  |  %s: %s  |  %s: %sx%s  |  ' 
                                                       %( __localize__(32141), get_language(item.get('iso_639_1','n/a')).capitalize(),
                                                          __localize__(32142), rating,
                                                          __localize__(32143), votes,
                                                          __localize__(32145), item.get('width'), item.get('height')))})
            except Exception, e:
                log( 'Problem report: %s' %str( e ), xbmc.LOGNOTICE )
            # Get posters
            try:
                for item in data['posters']:
                    if int(item.get('vote_count')) >= 1:
                        rating = float( "%.1f" % float( item.get('vote_average'))) #output string with one decimal
                        votes = item.get('vote_count','n/a')
                    else:
                        rating = 'n/a'
                        votes = 'n/a'
                    # Fill list
                    image_list.append({'url': BASE_IMAGEURL + 'original' + item['file_path'],
                                       'preview': BASE_IMAGEURL + 'w185' + item['file_path'],
                                       'id': item.get('file_path').lstrip('/').replace('.jpg', ''),
                                       'art_type': ['poster'],
                                       'height': item.get('height'),
                                       'width': item.get('width'),
                                       'language': item.get('iso_639_1','n/a'),
                                       'rating': rating,
                                       'votes': votes,
                                       # Create Gui string to display
                                       'generalinfo': ('%s: %s  |  %s: %s  |  %s: %s  |  %s: %sx%s  |  ' 
                                                       %( __localize__(32141), get_language(item.get('iso_639_1','n/a')).capitalize(),
                                                          __localize__(32142), rating,
                                                          __localize__(32143), votes,
                                                          __localize__(32145), item.get('width'), item.get('height')))})
            except Exception, e:
                log( 'Problem report: %s' %str( e ), xbmc.LOGNOTICE )
            if image_list == []:
                raise NoFanartError(media_id)
            else:
                # Sort the list before return. Last sort method is primary
                image_list = sorted(image_list, key=itemgetter('rating'), reverse=True)
                image_list = sorted(image_list, key=itemgetter('language'))
                return image_list


def _search_movie(medianame,year=''):
    medianame = normalize_string(medianame)
    log('TMDB API search criteria: Title[''%s''] | Year[''%s'']' % (medianame,year) )
    illegal_char = ' -<>:"/\|?*%'
    for char in illegal_char:
        medianame = medianame.replace( char , '+' ).replace( '++', '+' ).replace( '+++', '+' )

    # -JB-
    # (I'm looking for the year) OR (I'm looking for the year-1) OR (I'm looking for the year+1)
    for year_delta in [year, year-1,year+1]:
        search_url = 'http://api.themoviedb.org/3/search/movie?query=%s&primary_release_year=%s&api_key=%s' % (medianame, year_delta, API_KEY)

        tmdb_id = ''
        log('TMDB API search:   %s ' % search_url)
        try:
            data = get_data(search_url, 'json')
            if data == "Empty":
                tmdb_id = ''
            else:
                for item in data['results']:
                    if item['id']:
                        tmdb_id = item['id']
                        break
            if tmdb_id != '':
                break
        except Exception, e:
            log( str( e ), xbmc.LOGERROR )
    if tmdb_id == '':
        log('TMDB API search found no ID')
    else:
        log('TMDB API search found ID: %s' %tmdb_id)
    return tmdb_id

lib/provider/fanarttv.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
#     Copyright (C) 2011-2014 Martijn Kaijser
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program. If not, see <http://www.gnu.org/licenses/>.
#

#import modules
import sys
import urllib

### import libraries
#from lib.provider.base import BaseProvider
from lib.language import *
from lib.script_exceptions import NoFanartError
from lib.utils import *
from lib.settings import get
from operator import itemgetter

### get addon info
__localize__    = ( sys.modules[ '__main__' ].__localize__ )
setting = get()

API_KEY = '586118be1ac673f74963cc284d46bd8e'
API_URL_TV = 'http://webservice.fanart.tv/v3/tv/%s?api_key=%s'
API_URL_MOVIE = 'http://webservice.fanart.tv/v3/movies/%s?api_key=%s'
FANART_API_KEY = setting['fanart_api_key']
if FANART_API_KEY:
    API_KEY = FANART_API_KEY
    API_URL_TV = API_URL_TV + "&fanart_api_key=" + FANART_API_KEY
    API_URL_MOVIE = API_URL_MOVIE + "&fanart_api_key=" + FANART_API_KEY

IMAGE_TYPES_MOVIES = ['clearlogo',
                      'clearart',
                      'hdclearart',
                      'movielogo',
                      'hdmovielogo',
                      'movieart',
                      'moviedisc',
                      'hdmovieclearart',
                      'moviethumb',
                      'moviebanner']

IMAGE_TYPES_SERIES = ['clearlogo',
                      'hdtvlogo',
                      'clearart',
                      'hdclearart',
                      'tvthumb',
                      'seasonthumb',
                      'characterart',
                      'tvbanner',
                      'seasonbanner']

class FTV_TVProvider():

    def __init__(self):
        self.name = 'fanart.tv - TV API'

    def get_image_list(self, media_id):
        data = get_data(API_URL_TV%(media_id, API_KEY), 'json')
        image_list = []
        if data == 'Empty' or not data:
            return image_list
        else:
            # split 'name' and 'data'
            for value in data.iteritems():
                for art in IMAGE_TYPES_SERIES:
                    if art == value[0]:
                        for item in value[1]:
                            # Check on what type and use the general tag
                            arttypes = {'clearlogo': 'clearlogo',
                                        'hdtvlogo': 'clearlogo',
                                        'clearart': 'clearart',
                                        'hdclearart': 'clearart',
                                        'tvthumb': 'landscape',
                                        'seasonthumb': 'seasonlandscape',
                                        'characterart': 'characterart',
                                        'tvbanner': 'banner',
                                        'seasonbanner': 'seasonbanner',
                                        }
                            if art in ['hdtvlogo', 'hdclearart']:
                                size = 'HD'
                            elif art in ['clearlogo', 'clearart']:
                                size = 'SD'
                            else:
                                size = ''
                            # Create GUI info tag
                            generalinfo = '%s: %s  |  ' %( __localize__(32141), get_language(item.get('lang')).capitalize())
                            if item.get('season'):
                                generalinfo += '%s: %s  |  ' %( __localize__(32144), item.get('season'))
                            generalinfo += '%s: %s  |  ' %( __localize__(32143), item.get('likes'))
                            if art in ['hdtvlogo', 'hdclearart', 'clearlogo', 'clearart']:
                                generalinfo += '%s: %s  |  ' %( __localize__(32145), size)
                            # Fill list
                            image_list.append({'url': urllib.quote(item.get('url'), ':/'),
                                               'preview': item.get('url') + '/preview',
                                               'id': item.get('id'),
                                               'art_type': [arttypes[art]],
                                               'size': size,
                                               'season': item.get('season','n/a'),
                                               'language': item.get('lang'),
                                               'votes': int(item.get('likes')),
                                               'generalinfo': generalinfo})
            if image_list == []:
                raise NoFanartError(media_id)
            else:
                # Sort the list before return. Last sort method is primary
                image_list = sorted(image_list, key=itemgetter('votes'), reverse=True)
                image_list = sorted(image_list, key=itemgetter('size'), reverse=False)
                image_list = sorted(image_list, key=itemgetter('language'))
                return image_list
            
class FTV_MovieProvider():

    def __init__(self):
        self.name = 'fanart.tv - Movie API'

    def get_image_list(self, media_id):
        data = get_data(API_URL_MOVIE%(media_id, API_KEY), 'json')
        image_list = []
        if data == 'Empty' or not data:
            return image_list
        else:
            for value in data.iteritems():
                for art in IMAGE_TYPES_MOVIES:
                    if art == value[0]:
                        for item in value[1]:
                            # Check on what type and use the general tag
                            arttypes = {'movielogo': 'clearlogo',
                                        'moviedisc': 'discart',
                                        'movieart': 'clearart',
                                        'hdmovielogo': 'clearlogo',
                                        'hdmovieclearart': 'clearart',
                                        'moviebanner': 'banner',
                                        'moviethumb': 'landscape'}
                            if art in ['hdmovielogo', 'hdmovieclearart']:
                                size = 'HD'
                            elif art in ['movielogo', 'movieart']:
                                size = 'SD'
                            else:
                                size = ''
                            # Create GUI info tag
                            generalinfo = '%s: %s  |  ' %( __localize__(32141), get_language(item.get('lang')).capitalize())
                            if item.get('disc_type'):
                                generalinfo += '%s: %s (%s)  |  ' %( __localize__(32146), item.get('disc'), item.get('disc_type'))
                            if art in ['hdmovielogo', 'hdmovieclearart', 'movielogo', 'movieclearart']:
                                generalinfo += '%s: %s  |  ' %( __localize__(32145), size)
                            generalinfo += '%s: %s  |  ' %( __localize__(32143), item.get('likes'))
                            # Fill list
                            image_list.append({'url': urllib.quote(item.get('url'), ':/'),
                                               'preview': item.get('url') + '/preview',
                                               'id': item.get('id'),
                                               'art_type': [arttypes[art]],
                                               'size': size,
                                               'season': item.get('season','n/a'),
                                               'language': item.get('lang'),
                                               'votes': int(item.get('likes')),
                                               'disctype': item.get('disc_type','n/a'),
                                               'discnumber': item.get('disc','n/a'),
                                               'generalinfo': generalinfo})
            if image_list == []:
                raise NoFanartError(media_id)
            else:
                # Sort the list before return. Last sort method is primary
                image_list = sorted(image_list, key=itemgetter('votes'), reverse=True)
                image_list = sorted(image_list, key=itemgetter('size'), reverse=False)
                image_list = sorted(image_list, key=itemgetter('language'))
                return image_list